### PR TITLE
Basic implementation of custom format strings for Text() function.

### PR DESCRIPTION
**Pull request by LeeParrishMSFT**
A very basic implementation for custom string formats for the Text() function. 

Changes to the .csproj file were so that it's possible to add these projects to other solutions and make them project references to allow for editing and debugging across other projects that depend on these. 

Also added one small change to the LibraryDate.cs to get one more test to pass that was part of a Text() test.

### CODE REVIEW

**Comment by bryancroteau-MSFT** [comment]
Testing permissions to comment.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryDate.cs [hunk 1]
```diff
@@ -324,6 +324,11 @@ public static FormulaValue Date(IRContext irContext, NumberValue[] args)
             var month = (int)args[1].Value;
             var day = (int)args[2].Value;
 
+            if (year <= 1899)
+            {
+                year = 1900 + year;
```
Comment (bryancroteau-MSFT) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryDate.cs:329]: I don't think we plan to support this.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs [hunk 2]
```diff
@@ -115,45 +115,47 @@ public static FormulaValue Value(EvalVisitor runner, SymbolContext symbolContext
...
-            else if (arg0 is DateValue d)
-            {
-                // $$$ Use real format string
-                str = d.Value.ToString("M/d/yyyy", runner.CultureInfo);
-            }
-            else if (arg0 is DateTimeValue dt)
+            string resultString = null;
+            string formatString = null;
+
+            if (args.Length > 1 && args[1] is StringValue)
```
Comment (bryancroteau-MSFT) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs:127]: args[1] is StringValue **sv**

Use the variable **sv** in the body of the if statement.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Text.txt [hunk 3]
```diff
@@ -27,39 +26,31 @@
...
 "123,456,789.0"
 
 >> Text(123456789, "0,00.0")
-/*InterpreterRunner*/ #Skip
 "123,456,789.0"
 
 >> Text(123456789, "0,")
-/*InterpreterRunner*/ #Skip
-"123456789,"
+"123457"
```
Comment (bryancroteau-MSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Text.txt:47]: Has this been tested in the TypeScript implementation?

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/FileExpressionEvaluationTests.cs [hunk 4]
```diff
@@ -79,6 +79,10 @@ internal void TestToString(FormulaValue result, StringBuilder sb)
             {
                 sb.Append(n.Value);
             }
+            else if (result is DateValue d)
```
Comment (bryancroteau-MSFT) [inline src/tests/Microsoft.PowerFx.Interpreter.Tests/FileExpressionEvaluationTests.cs:82]: Do we have DateTimeValue and TimeValue branches for this as well?

**Review by bryancroteau-MSFT — CHANGES_REQUESTED** [review]

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Text.txt (hunk 3 above)
Comment (LeeParrishMSFT) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Text.txt:47]: I'll have to get with you on how to test against the TypeScript implementation as I'm not sure how to do that. I did check against Excel and this is what it returns when provided the same input.

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/FileExpressionEvaluationTests.cs (hunk 4 above)
Comment (LeeParrishMSFT) [reply src/tests/Microsoft.PowerFx.Interpreter.Tests/FileExpressionEvaluationTests.cs:82]: I only added what I need to pass existing tests that were failing. I'm guessing there may be needs for DateTimeValue and TimeValue at some point, but I haven't seen tests that require these yet.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryDate.cs (hunk 1 above)
Comment (LeeParrishMSFT) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryDate.cs:329]: I'm fine with that, but if we don't then we will need to remove the test that checks this quirk:

>> Date(80, 1, 2)
1/2/1980

Comment (bryancroteau-MSFT) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryDate.cs:329]: This test should still work in the TypeScript backend for now. We can just skip that test or provide an alternative answer for just the C# interpreter.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Text.txt (hunk 3 above)
Comment (bryancroteau-MSFT) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Text.txt:47]: I wonder why Excel does it this way. You can also use the prod PowerFx client to test by hand.

Comment (bryancroteau-MSFT) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Text.txt:47]: I just checked in Canvas Apps and the original test was correct. We can ask Greg why this is the case.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryDate.cs (hunk 1 above)
Comment (LeeParrishMSFT) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryDate.cs:329]: I've removed the change and modified the test to have an expected output that matches with the C# implementation.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Text.txt (hunk 3 above)
Comment (LeeParrishMSFT) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Text.txt:47]: I've added an expected result for the InterpreterRunner since this is the value we expect based on Excel. If TypeScript is different then they need to work out why, but the base expected result is now the original "123456789," value.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs [hunk 5]
```diff
@@ -115,45 +115,47 @@ public static FormulaValue Value(EvalVisitor runner, SymbolContext symbolContext
...
+                    resultString = s.Value;
+                    break;
+                case DateValue d:
+                    resultString = d.Value.ToString(formatString ?? "M/d/yyyy", runner.CultureInfo);
+                    break;
+                case DateTimeValue dt:
+                    resultString = dt.Value.ToString(formatString ?? "g", runner.CultureInfo);
+                    break;
+                case TimeValue t:
+                    resultString = new DateTime().Add(t.Value).ToString(formatString ?? "t", runner.CultureInfo);
```
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs:147]: Low priority, but if the format contains a year, this will be different than Excel: `Text(Time(13,35,57), "yyyy-mm hh:mm:ss")` - it will show "1900-01 13:35:57" in Excel, and "0001-01 13:35:57" with this implementation. We should align with the "0-value" for the date (1899-12-31)

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Text.txt [hunk 6]
```diff
@@ -81,4 +72,16 @@ Blank()
...
\ No newline at end of file
+#Error
+
+>> Text(DateTimeValue("03/15/2018 12:00:00"), "yyyy-MM-ddTHH:mm:ss")
+"2018-03-15T12:00:00"
+
+>> Text(DateValue("03/15/2018 12:00:00"), "yyyy-MM-dd")
+"2018-03-15"
+
+>> Text(DateValue("03/15/2018"), "yyyy-MM-dd")
```
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Text.txt:84]: It is better to use `Date(2018, 3, 15)` instead of `DateValue("03/15/2018")` since the former works in all regions, and the latter won't work in regions where the date format is dd/mm/yyyy.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Text.txt [hunk 7]
```diff
@@ -81,4 +72,16 @@ Blank()
 "B"
 
 >> Char(1/0)
-#Error
\ No newline at end of file
+#Error
+
+>> Text(DateTimeValue("03/15/2018 12:00:00"), "yyyy-MM-ddTHH:mm:ss")
```
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Text.txt:78]: Similar to the comment below. We should use something like `Date(2018, 3, 15) + Time(12, 0, 0)`; check with @bryancroteau-MSFT if this is already implemented, though.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs (hunk 5 above)
Comment (bryancroteau-MSFT) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs:147]: Right, raw times without a date will be relative to the epoch defined in CoercionUnary.cs

Comment (bryancroteau-MSFT) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs:147]: So this should really be

`_epoch.Add(t.Value)...`

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Text.txt (hunk 7 above)
Comment (bryancroteau-MSFT) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Text.txt]: This is implemented in LibraryOperators.cs

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Text.txt (hunk 6 above)
Comment (LeeParrishMSFT) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Text.txt]: Got it, thanks!

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Text.txt (hunk 7 above)
Comment (LeeParrishMSFT) [reply src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/Text.txt]: I've removed this test as I added it to check something I need for a specific purpose and having the test re-written as shown doesn't really prove what I needed so the purpose of the test is lost.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs (hunk 5 above)
Comment (LeeParrishMSFT) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs:147]: Getting the start value to be in alignment is one issue, and we can fix that pretty easily with a "new DateTime(1900,1,1).Add(t.Value)...  The bigger issues come with the .net runtime is there are differences between "mm" and "MM" and "hh" and "HH". The .net runtime interprets the format string as wanting us to return "mm" or minutes and "hh" as in 12 hour clock hours, so you get back is: "1900-35 01:35:57" and to the .net runtime this is exactly what we asked for based on a format string of "yyyy-mm hh:mm:ss".  We would have to start modifying format strings to replicate the behavior of Excel which is what we were trying to avoid on this simple implementation. Once we get a comprehensive test set that can exercise all these quirks, then we can start to address them knowing we aren't breaking one thing to fix something else.

Comment (bryancroteau-MSFT) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs]: The epoch for our implementation is actually 1899-12-30 as implied mentioned above. This is because Excel incorrectly thinks that the year 1900 is a leap year, but it is not. So our epoch is offset by one day for negative numbers, 0, and the first 60 or so days.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs [hunk 8]
```diff
@@ -115,45 +115,47 @@ public static FormulaValue Value(EvalVisitor runner, SymbolContext symbolContext
...
+                    resultString = s.Value;
+                    break;
+                case DateValue d:
+                    resultString = d.Value.ToString(formatString ?? "M/d/yyyy", runner.CultureInfo);
+                    break;
+                case DateTimeValue dt:
+                    resultString = dt.Value.ToString(formatString ?? "g", runner.CultureInfo);
+                    break;
+                case TimeValue t:
+                    resultString = new DateTime(1900, 1, 1, 0, 0, 0, DateTimeKind.Utc).Add(t.Value).ToString(formatString ?? "t", runner.CultureInfo);
```
Comment (bryancroteau-MSFT) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs:147]: As I mentioned our actual epoch is not the same as Excel.

**Review by bryancroteau-MSFT — CHANGES_REQUESTED** [review]

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs (hunk 8 above)
Comment (LeeParrishMSFT) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryText.cs]: I've made the change to use _epoch rather than DateTime(1900, etc.)

**Review by bryancroteau-MSFT — APPROVED** [review]
Ship it

File: src/libraries/Microsoft.PowerFx.Core/Microsoft.PowerFx.Core.csproj [hunk 9]
```diff
@@ -29,10 +29,10 @@
 	</ItemGroup>
 
 	<ItemGroup>
-		<EmbeddedResource Include="$(SolutionDir)\strings\**\PowerFxResources.resw" Condition="'$(Configuration)' == 'Debug'">
+		<EmbeddedResource Include="..\..\strings\**\PowerFxResources.resw" Condition="'$(Configuration)' == 'Debug'">
```
Comment (lesaltzm) [inline src/libraries/Microsoft.PowerFx.Core/Microsoft.PowerFx.Core.csproj:32]: Why this change?

Comment (lesaltzm) [reply src/libraries/Microsoft.PowerFx.Core/Microsoft.PowerFx.Core.csproj:32]: it has no real impact, just curious.

**Review by lesaltzm — APPROVED** [review]
:ship:

File: src/libraries/Microsoft.PowerFx.Core/Microsoft.PowerFx.Core.csproj (hunk 9 above)
Comment (LeeParrishMSFT) [reply src/libraries/Microsoft.PowerFx.Core/Microsoft.PowerFx.Core.csproj:32]: When I'm working on the code to convert from Adaptive to Pfx I add the Pfx projects to the Conversion solution and then edit/build/debug across both code bases which allows me to work on both at the same time. But having $(SolutionDir)\ as the path for the string resources doesn't work if you add the Pfx projects to a different solution and the solution isn't in the same folder tree. Very esoteric, but as you said, it really doesn't change anything on the Pfx side in that today both are equivalent anyway.
